libvast/src/type.cpp

```diff
@@ -2582,4 +3201,65 @@ sum_type_access<vast::type>::index_from_type(const vast::type& x) noexcept {
   return result;
 }
 
+int sum_type_access<arrow::DataType>::index_from_type(
+  const arrow::DataType& x) noexcept {
+  using extension_types = detail::tl_filter_t<types, arrow::is_extension_type>;
+  static constexpr int extension_id = -1;
+  static constexpr int unknown_id = -2;
+  // The first-stage O(1) lookup table from arrow::DataType id to the sum type
+  // variant index defined by the type list. Returns unknown_id if the DataType
+  // is not in the type list, and extension_id if the type is an extension type.
+  static const auto table = []<class... Ts, int... Indices>(
+    caf::detail::type_list<Ts...>,
+    std::integer_sequence<int, Indices...>) noexcept {
+    std::array<int, arrow::Type::type::MAX_ID> tbl{};
+    tbl.fill(unknown_id);
+    (static_cast<void>(tbl[Ts::type_id] = arrow::is_extension_type<Ts>::value
+                                            ? extension_id
+                                            : Indices),
+     ...);
+    return tbl;
+  }
+  (sum_type_access<arrow::DataType>::types{},
+   std::make_integer_sequence<int, caf::detail::tl_size<types>::value>());
+  // The second-stage O(n) lookup table for extension types that identifies the
+  // types by their unique identifier string.
+  static const auto extension_table = []<class... Ts, int... Indices>(
+    caf::detail::type_list<Ts...>, std::integer_sequence<int, Indices...>) {
+    std::array<std::pair<std::string_view, int>,
+               detail::tl_size<extension_types>::value>
+      tbl{};
+    (static_cast<void>(
+       tbl[Indices]
+       = {Ts::name, detail::tl_index_of<sum_type_access<arrow::DataType>::types,
+                                        Ts>::value}),
+     ...);
+    return tbl;
+  }
+  (extension_types{},
+   std::make_integer_sequence<int,
+                              caf::detail::tl_size<extension_types>::value>());
+  // NOLINTNEXTLINE(cppcoreguidelines-pro-bounds-constant-array-index)
+  auto result = table[x.id()];
+  VAST_ASSERT(result != unknown_id,
+              "unexpected Arrow type id is not in "
+              "caf::sum_type_access<arrow::DataType>::types");
+  if (result == extension_id) {
+    for (const auto& [id, index] : extension_table) {
+      if (id == static_cast<const arrow::ExtensionType&>(x).extension_name())
+        return index;
+    }
+    vast::die("unexpected Arrow extension type");
+  }
+  return result;
+}
+
+/// Explicit template instantiations for all defined sum type access
+/// specializations.
+template struct sum_type_access<vast::type>;
+template struct sum_type_access<arrow::DataType>;
+template struct sum_type_access<arrow::Array>;
+template struct sum_type_access<arrow::Scalar>;
+template struct sum_type_access<arrow::ArrayBuilder>;
```

Comment:
I believe these aren't actually "extern" unless you move the member function definitions to this cpp file.
At least, I can build and run the tests without these lines.